### PR TITLE
rgw: validate rgw endpoint ip

### PR DESCRIPTION
**Pull request by parth-gr**
The ceph-external-cluster-details-exporter.py
generates the json output even with
rgw-endpoint IP from different cluster
Added a check validate_rgw_endpoint which will validate
the above rgw_endpoint Ip by matching the cluster fsid

Signed-off-by: parth-gr <paarora@redhat.com>

**Description of your changes:**

**Which issue is resolved by this Pull Request:**
Resolves #

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: If this is only a documentation change, add the label `skip-ci` on the PR.
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/development-flow.html#submitting-a-pull-request)
- [ ] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

### CODE REVIEW

File: deploy/examples/create-external-cluster-resources.py [hunk 1]
```diff
@@ -213,7 +216,7 @@ def gen_arg_parser(cls, args_to_parse=None):
 
         upgrade_group = argP.add_argument_group('upgrade')
         upgrade_group.add_argument("--upgrade", action='store_true', default=False,
-                                   help="Upgrades the 'csi-user'(For example: client.csi-cephfs-provisioner) with new permissions needed for the new cluster version and older permission will still be applied." +
+                                   help="Upgrades the cephCSIKeyrings(For example: client.csi-cephfs-provisioner) with new permissions needed for the new cluster version and older permission will still be applied." +
```
Comment (subhamkrai) [inline deploy/examples/create-external-cluster-resources.py:219]:
```suggestion
                                   help="Upgrades the 'csi-user'(For example: client.csi-cephfs-provisioner) with new permissions needed for the new cluster version and older permission will still be applied." +
```
seems good to me as it csi ceph user

File: deploy/examples/create-external-cluster-resources.py [hunk 2]
```diff
@@ -826,7 +829,44 @@ def validate_rados_namespace(self):
...
+        http_request = "GET"
+        rgw_endpoint = self._arg_parser.rgw_endpoint
+        resource = "/admin/info"
+        query_string = "format=json"
+        contentType = "application/x-compressed-tar"
+        dateTime = datetime.datetime.utcnow()
+        dateTimeFormat = dateTime.strftime('%a, %d %b %Y %X +0000')
+        headerToSign = http_request + "\n" + "\n" + contentType + "\n" + dateTimeFormat + "\n" + resource
+        print(headerToSign)
+        cmd = "echo -en " + headerToSign + " | openssl sha1 -hmac " + secret_key +" -binary | base64"
```
Comment (leseb) [inline deploy/examples/create-external-cluster-resources.py:847]: let's be careful if `openssl` does not exist on the machine. Alternatively, it will be good to see Python has something we can use as a lib instead.

File: deploy/examples/create-external-cluster-resources.py [hunk 3]
```diff
@@ -826,7 +829,44 @@ def validate_rados_namespace(self):
             raise ExecutionFailureException(
                 ("The provided rados Namespace, '{}', is not found in the pool '{}'").format(
                     rados_namespace, rbd_pool_name))
-
+    
+    def get_rgw_fsid(self):
+        access_key = self.out_map['RGW_ADMIN_OPS_USER_ACCESS_KEY']
+        access_key = "WC2D69F9L7KEOJVETKER"
```
Comment (thotz) [inline deploy/examples/create-external-cluster-resources.py:835]: why access key and secret key have hardcoded values?

File: deploy/examples/create-external-cluster-resources.py [hunk 4]
```diff
@@ -826,7 +829,44 @@ def validate_rados_namespace(self):
...
+        print(headerToSign)
+        cmd = "echo -en " + headerToSign + " | openssl sha1 -hmac " + secret_key +" -binary | base64"
+        try:
+            signature_decode = subprocess.Popen(cmd,stderr=subprocess.PIPE,stdout=subprocess.PIPE,shell=True,universal_newlines=True)
+            signature = signature_decode.communicate()[0].strip()
+        except subprocess.CalledProcessError as execErr:
+            err_msg = "failed to execute command %s. Output: %s. Code: %s. Error: %s" % (
+                cmd, execErr.output, execErr.returncode, execErr.stderr)
+            raise Exception(err_msg)
+        request_url = "http://" + rgw_endpoint + resource + "?" + query_string
```
Comment (thotz) [inline deploy/examples/create-external-cluster-resources.py:855]: what if RGW is running in htttps?

File: deploy/examples/create-external-cluster-resources.py [hunk 5]
```diff
@@ -768,7 +771,7 @@ def get_ceph_dashboard_link(self):
 
     def create_rgw_admin_ops_user(self):
         cmd = ['radosgw-admin', 'user', 'create', '--uid', self.EXTERNAL_RGW_ADMIN_OPS_USER_NAME, '--display-name',
-               'Rook RGW Admin Ops user', '--caps', 'buckets=*;users=*;usage=read;metadata=read;zone=read']
+               'Rook RGW Admin Ops user', '--caps', 'info=read;buckets=*;users=*;usage=read;metadata=read;zone=read']
```
Comment (thotz) [inline deploy/examples/create-external-cluster-resources.py:774]: @leseb what about adding `info=read` for the admins ops user created by Rook to make it consistent

Comment (leseb) [reply deploy/examples/create-external-cluster-resources.py:774]: That wouldn't hurt for sure but go-ceph does not support it yet. Perhaps we can do that in another PR once it's supported.

Comment (thotz) [reply deploy/examples/create-external-cluster-resources.py:774]: As far as I understand go-ceph handles the` user_cap` struct as a generic struct containing `type` and `perm` as strings. So IMO no change is required from the go-ceph library and we can directly update the perms for `admin ops user`. On a different note if u need to include perms for `cephobjectstoreuser` then we need some changes for its Spec

Comment (leseb) [reply deploy/examples/create-external-cluster-resources.py:774]: I meant handling the call to `info/` path.

File: deploy/examples/create-external-cluster-resources.py (hunk 1 above)
Comment (parth-gr) [reply deploy/examples/create-external-cluster-resources.py:219]: It's actually the keyring only, the user is created below with the help of this keyring.

File: deploy/examples/create-external-cluster-resources.py (hunk 3 above)
Comment (parth-gr) [reply deploy/examples/create-external-cluster-resources.py]: Yaa removed it, added mistakenly test cluster info

File: deploy/examples/create-external-cluster-resources.py (hunk 4 above)
Comment (parth-gr) [reply deploy/examples/create-external-cluster-resources.py:853]: @thotz I am not sure about it,
If rgw endpoint can be configured with both ways, how we can detect it,
Or we should ask from user about it as a cli parameter ?

File: deploy/examples/create-external-cluster-resources.py (hunk 5 above)
Comment (thotz) [reply deploy/examples/create-external-cluster-resources.py:774]: okay

File: deploy/examples/create-external-cluster-resources.py (hunk 4 above)
Comment (thotz) [reply deploy/examples/create-external-cluster-resources.py:853]: @parth-gr : AFAIR the script has parameter to TLS certs so that it will create k8s out of it. If it is passed then  we can assume it is `https` otherwise http

File: deploy/examples/create-external-cluster-resources.py [hunk 6]
```diff
@@ -826,7 +955,27 @@ def validate_rados_namespace(self):
...
-
+    
+    def get_rgw_fsid(self):
+        access_key = self.out_map['RGW_ADMIN_OPS_USER_ACCESS_KEY']
+        secret_key = self.out_map['RGW_ADMIN_OPS_USER_SECRET_KEY']
+        rgw_endpoint = self._arg_parser.rgw_endpoint
+        request_url = "http://"
+        if self._arg_parser.rgw_tls_cert_path:
+            request_url = "https://"
+        request_url = request_url + rgw_endpoint + "/admin/info" + "?" + "format=json"
```
Comment (thotz) [inline deploy/examples/create-external-cluster-resources.py:966]: I am assuming this curl request sending to rgw_endpoint, so if it is `HTTPS` then either need to attach certificate or use `-k` for insecure flag for curl command. Otherwise request will fail

**Review by thotz — APPROVED** [review]
Overall code workflow looks good to me, one minor comment from me. I don't know in-depth python knowledge to check code validity

File: .github/workflows/canary-integration-test.yml [hunk 7]
```diff
@@ -164,6 +164,64 @@ jobs:
...
+    runs-on: ubuntu-18.04
+    if: "!contains(github.event.pull_request.labels.*.name, 'skip-ci')"
+    steps:
+    - name: checkout
+      uses: actions/checkout@v2
+      with:
+        fetch-depth: 0
+
+    - name: update ceph to quincy
+      run:  export CEPH_QUINCY=true
```
Comment (leseb) [inline .github/workflows/canary-integration-test.yml:177]:
```suggestion
      run: export CEPH_QUINCY=true
```

File: images/ceph/Makefile [hunk 8]
```diff
@@ -22,6 +22,11 @@ CEPH_VERSION = v16.2.7-20220216
 else
 CEPH_VERSION = v16.2.7-20220216
 endif
+
+ifeq ($(CEPH_QUINCY),true)
```
Comment (leseb) [inline images/ceph/Makefile:26]: Why not override `BASEIMAGE` instead of adding a new env var?

File: .github/workflows/canary-integration-test.yml [hunk 9]
```diff
@@ -164,6 +164,64 @@ jobs:
...
+        mgr_raw=$(kubectl -n rook-ceph exec $toolbox -- ceph mgr dump -f json|jq --raw-output .active_addr)
+        timeout 60 sh -c "until kubectl -n rook-ceph exec $toolbox -- curl --silent --show-error ${mgr_raw%%:*}:9283; do echo 'waiting for mgr prometheus exporter to be ready' && sleep 1; done"
+        kubectl -n rook-ceph cp deploy/examples/create-external-cluster-resources.py $toolbox:/etc/ceph
+        # pass the valid rgw-endpoint of same ceph cluster
+        timeout 60 sh -c "until kubectl -n rook-ceph exec $toolbox -- python3 /etc/ceph/create-external-cluster-resources.py --rbd-data-pool-name replicapool --rgw-endpoint $rgw_endpoint; do sleep 1 && echo 'waiting for the rgw endpoint to be validated'; done"
+        # pass the invalid rgw-endpoint of different ceph cluster
+        timeout 60 sh -c "until kubectl -n rook-ceph exec $toolbox -- python3 /etc/ceph/create-external-cluster-resources.py --rbd-data-pool-name replicapool --rgw-endpoint "10.96.77.46:80"; do sleep 1 && echo 'waiting for the rgw endpoint to be validated'; done"
```
Comment (leseb) [inline .github/workflows/canary-integration-test.yml:212]: this is expected to fail right? so we need to catch the return code and ignore it if it's the one we are looking for.

File: deploy/examples/create-external-cluster-resources.py [hunk 10]
```diff
@@ -213,7 +342,7 @@ def gen_arg_parser(cls, args_to_parse=None):
 
         upgrade_group = argP.add_argument_group('upgrade')
         upgrade_group.add_argument("--upgrade", action='store_true', default=False,
-                                   help="Upgrades the 'csi-user'(For example: client.csi-cephfs-provisioner) with new permissions needed for the new cluster version and older permission will still be applied." +
+                                   help="Upgrades the cephCSIKeyrings(For example: client.csi-cephfs-provisioner) with new permissions needed for the new cluster version and older permission will still be applied." +
```
Comment (leseb) [inline deploy/examples/create-external-cluster-resources.py:345]: let this be a separate commit

File: deploy/examples/create-external-cluster-resources.py [hunk 11]
```diff
@@ -826,7 +954,27 @@ def validate_rados_namespace(self):
...
-
+    
+    def get_rgw_fsid(self):
+        access_key = self.out_map['RGW_ADMIN_OPS_USER_ACCESS_KEY']
+        secret_key = self.out_map['RGW_ADMIN_OPS_USER_SECRET_KEY']
+        rgw_endpoint = self._arg_parser.rgw_endpoint
+        request_url = "http://"
+        if self._arg_parser.rgw_tls_cert_path:
+            request_url = "https://"
+        request_url = request_url + rgw_endpoint + "/admin/info" + "?" + "format=json"
```
Comment (leseb) [inline deploy/examples/create-external-cluster-resources.py:965]: Doesn't Python have a lib to build URI? Instead of the "manual" string concat. The equivalent of go `url.Values{}` from `"net/url"`.

File: deploy/examples/create-external-cluster-resources.py [hunk 12]
```diff
@@ -826,7 +954,27 @@ def validate_rados_namespace(self):
...
+        access_key = self.out_map['RGW_ADMIN_OPS_USER_ACCESS_KEY']
+        secret_key = self.out_map['RGW_ADMIN_OPS_USER_SECRET_KEY']
+        rgw_endpoint = self._arg_parser.rgw_endpoint
+        request_url = "http://"
+        if self._arg_parser.rgw_tls_cert_path:
+            request_url = "https://"
+        request_url = request_url + rgw_endpoint + "/admin/info" + "?" + "format=json"    
+        r = requests.get(request_url, auth=S3Auth(access_key, secret_key,rgw_endpoint))
+        r1 = r.json()
+        return r1['info']['storage_backends'][0]['cluster_id']
```
Comment (leseb) [inline deploy/examples/create-external-cluster-resources.py:968]: verify the return is not `null`?

**Review by leseb — CHANGES_REQUESTED** [review]

File: images/ceph/Makefile (hunk 8 above)
Comment (leseb) [reply images/ceph/Makefile:26]: You don't need to export, you can do `BASEIMAGE=foo make build`

File: .github/workflows/canary-integration-test.yml [hunk 13]
```diff
@@ -164,6 +164,64 @@ jobs:
...
+      with:
+        fetch-depth: 0
+
+    - name: update ceph to quincy
+      run: export CEPH_QUINCY=true
+    
+    - name: setup cluster resources
+      uses: ./.github/workflows/setup-cluster-resources
+      with:
+        github-token: ${{ secrets.GITHUB_TOKEN }}
```
Comment (BlaineEXE) [inline .github/workflows/canary-integration-test.yml:182]: I think this might need `env`

```yaml
env:
  CEPH_QUINCY=true
``

File: images/ceph/Makefile [hunk 14]
```diff
@@ -22,6 +22,13 @@ CEPH_VERSION = v16.2.7-20220216
...
+CEPH_VERSION = v17.2.0-20220420
+endif
+
+
+
```
Comment (BlaineEXE) [inline images/ceph/Makefile:31]: This is really unnecessary if we just make the CEPH_VERSION var overrideably by changing `CEPH_VERSION = <something>` to `CEPH_VERSION ?= <default-value>`.

File: deploy/examples/create-external-cluster-resources.py (hunk 6 above)
Comment (parth-gr) [reply deploy/examples/create-external-cluster-resources.py]: I am not sure about networking, 
@leseb can you assist with what else to do for HTTPS request/

File: .github/workflows/canary-integration-test.yml (hunk 9 above)
Comment (parth-gr) [reply .github/workflows/canary-integration-test.yml]: I just removed it, as if it fails the CI will not run successfully, 
Or maybe any other way you may tell, or else there is no need for false testing.

File: deploy/examples/create-external-cluster-resources.py [hunk 15]
```diff
@@ -149,6 +166,118 @@ def _convert_hostname_to_ip(self, host_name):
     def Rados(conffile=None):
         return DummyRados()
 
+class S3Auth(AuthBase):
+
+    """Attaches AWS Authentication to the given Request object."""
+
+    service_base_url = 's3.amazonaws.com'
+    # List of Query String Arguments of Interest
+    special_params = [
```
[the remark this answers is not in the document]
Comment (parth-gr) [reply deploy/examples/create-external-cluster-resources.py]: Removed

File: deploy/examples/create-external-cluster-resources.py (hunk 6 above)
Comment (leseb) [reply deploy/examples/create-external-cluster-resources.py]: It depends, where is HTTP client we build to run the request?

Comment (parth-gr) [reply deploy/examples/create-external-cluster-resources.py]: OKay, the URI is made like this only  `  request_url = request_url + rgw_endpoint + "/admin/info" + "?" + "format=json"   `

File: .github/workflows/canary-integration-test.yml (hunk 9 above)
Comment (parth-gr) [reply .github/workflows/canary-integration-test.yml]: updated,thanks

**Review by leseb — CHANGES_REQUESTED** [review]

File: deploy/examples/create-external-cluster-resources-tests.py [hunk 16]
```diff
@@ -34,8 +34,7 @@
 class TestRadosJSON(unittest.TestCase):
     def setUp(self):
         print("\nI am in setup")
-        self.rjObj = ext.RadosJSON(['--rbd-data-pool-name=abc',
-                                    '--rgw-endpoint=10.10.212.122:9000', '--format=json'])
+        self.rjObj = ext.RadosJSON(['--rbd-data-pool-name=abc', '--format=json'])
```
[the remark this answers is not in the document]
Comment (parth-gr) [reply deploy/examples/create-external-cluster-resources-tests.py:37]: This is because we are now making the endpoint call through admin ops API, and there should be something that should be returned from the server, and the above endpoint is just random numbers
So as I mentioned it is not mockable for unit tests and it fails if we keep otherwise.

File: deploy/examples/create-external-cluster-resources.py [hunk 17]
```diff
@@ -826,7 +938,37 @@ def validate_rados_namespace(self):
...
+        rgw_endpoint = self._arg_parser.rgw_endpoint
+        base_url = "http://"
+        if self._arg_parser.rgw_tls_cert_path:
+            base_url = "https://"
+        base_url = base_url + rgw_endpoint + "/admin/info?"   
+        params = {'format': 'json'}
+        request_url = base_url + urllib.parse.urlencode(params)
+            
+        try:
+            r = requests.get(request_url, auth=S3Auth(access_key, secret_key,rgw_endpoint))
```
Comment (leseb) [inline deploy/examples/create-external-cluster-resources.py:954]: Does `get()` accept a modified client? How can we pass the certificate or ignore the certificate verification?

File: deploy/examples/create-external-cluster-resources.py [hunk 18]
```diff
@@ -826,7 +938,37 @@ def validate_rados_namespace(self):
...
+            return "NULL"
+        
+        return r1['info']['storage_backends'][0]['cluster_id']
+            
+    def validate_rgw_endpoint(self):
+        fsid = self.get_fsid()
+        rgw_fsid = self.get_rgw_fsid()
+        if fsid != rgw_fsid:
+            raise ExecutionFailureException(
+                ("The provided rgw Endpoint, '{}', is invalid. We are validating by calling the adminops api through rgw-endpoint and checking the cluster_id '{}' we get is same as the fsid of the ceph cluster '{}'").format(
```
Comment (leseb) [inline deploy/examples/create-external-cluster-resources.py:969]: `checking the cluster_id '{}' we get is same as the` I don't think this is grammatically correct.

File: deploy/examples/create-external-cluster-resources.py (hunk 17 above)
Comment (parth-gr) [reply deploy/examples/create-external-cluster-resources.py]: Yes it can be modified,  changed some code let see if it passes the CI

Comment (parth-gr) [reply deploy/examples/create-external-cluster-resources.py]: @leseb The changes are working, just need to understand what will the cert value/path that I can provide in the CI

File: deploy/examples/create-external-cluster-resources.py [hunk 19]
```diff
@@ -826,7 +938,42 @@ def validate_rados_namespace(self):
...
+        base_url =  self.endpoint_dial(rgw_endpoint) + "://"
+        cert = None
+        if self._arg_parser.rgw_tls_cert_path:
+            cert = self._arg_parser.rgw_tls_cert_path
+        base_url = base_url + rgw_endpoint + "/admin/info?"   
+        params = {'format': 'json'}
+        request_url = base_url + urllib.parse.urlencode(params)
+            
+        try:
+            r = requests.get(request_url, auth=S3Auth(access_key, secret_key,rgw_endpoint), cert=cert)
```
Comment (leseb) [inline deploy/examples/create-external-cluster-resources.py:955]: Can `cert` be empty in `get()`? Is there any option to do insecure TLS?

Comment (parth-gr) [reply deploy/examples/create-external-cluster-resources.py:955]: Yes it can, added a check for `self.arg_parser.rgw_skip_tls == False` 

https://github.com/rook/rook/pull/10006#discussion_r861489844 can you plz give reply to it so I can verify too

File: deploy/examples/create-external-cluster-resources.py (hunk 17 above)
Comment (leseb) [reply deploy/examples/create-external-cluster-resources.py]: We cannot use certs with the CI since they are self-signed. We can however use the insecure TLS option if we deploy RGW with TLS in the CI.

File: deploy/examples/create-external-cluster-resources.py (hunk 19 above)
Comment (leseb) [reply deploy/examples/create-external-cluster-resources.py:955]: When `self.arg_parser.rgw_skip_tls == True` we **must** pass an option like insecure TLS.

Comment (parth-gr) [reply deploy/examples/create-external-cluster-resources.py:955]: these are the get parameter we can apply https://www.w3schools.com/python/ref_requests_get.asp,

In which we can apply `verify=false`

File: deploy/examples/create-external-cluster-resources.py (hunk 17 above)
Comment (parth-gr) [reply deploy/examples/create-external-cluster-resources.py]: I saw the Pr in which you added the TLS and gave the example of `--rgw-tls-cert-path my-cert` https://github.com/rook/rook/commit/d6451527437e8e999c3ea6d38d17f7969e32b989 so if some How I can replicate this behaviour.

File: deploy/examples/create-external-cluster-resources.py [hunk 20]
```diff
@@ -826,7 +938,42 @@ def validate_rados_namespace(self):
...
                 ("The provided rados Namespace, '{}', is not found in the pool '{}'").format(
                     rados_namespace, rbd_pool_name))
-
+    
+    def get_rgw_fsid(self):
+        access_key = self.out_map['RGW_ADMIN_OPS_USER_ACCESS_KEY']
+        secret_key = self.out_map['RGW_ADMIN_OPS_USER_SECRET_KEY']
+        rgw_endpoint = self._arg_parser.rgw_endpoint
+        cert = None
+        if self._arg_parser.rgw_tls_cert_path and self._arg_parser.rgw_skip_tls == False :
```
Comment (leseb) [inline deploy/examples/create-external-cluster-resources.py:947]:
```suggestion
        if self._arg_parser.rgw_tls_cert_path and not self._arg_parser.rgw_skip_tls:
```

???

File: deploy/examples/create-external-cluster-resources.py (hunk 19 above)
Comment (leseb) [reply deploy/examples/create-external-cluster-resources.py:955]: Yes, that's what we want when `rgw_skip_tls` is True.

File: deploy/examples/create-external-cluster-resources.py [hunk 21]
```diff
@@ -826,7 +938,46 @@ def validate_rados_namespace(self):
...
+        r1 = r.json()
+        if r1 is None or r1.get('info') is None:
+            return "NULL"
+        
+        return r1['info']['storage_backends'][0]['cluster_id']
+            
+    def validate_rgw_endpoint(self):
+        # if the 'cluster' instance is a dummy one,
+        # don't try to reach out to the endpoint
+        if isinstance(self.cluster, DummyRados):
```
Comment (leseb) [inline deploy/examples/create-external-cluster-resources.py:972]: That's unit test convenience, right? Can we remove it from the main code?

File: deploy/examples/create-external-cluster-resources.py [hunk 22]
```diff
@@ -768,7 +869,7 @@ def get_ceph_dashboard_link(self):
 
     def create_rgw_admin_ops_user(self):
         cmd = ['radosgw-admin', 'user', 'create', '--uid', self.EXTERNAL_RGW_ADMIN_OPS_USER_NAME, '--display-name',
-               'Rook RGW Admin Ops user', '--caps', 'buckets=*;users=*;usage=read;metadata=read;zone=read']
+               'Rook RGW Admin Ops user', '--caps', 'info=read;buckets=*;users=*;usage=read;metadata=read;zone=read']
```
Comment (leseb) [inline deploy/examples/create-external-cluster-resources.py:872]: I think we should only add this if the cluster version is at least Quincy otherwise the script will always fail.
We need to:

* Check the local binary version with `ceph -v`
* Check the cluster's version with `ceph version`

If not Qunicy, don't add the `info=read` and log a message that we won't be able to validate the rgw endpoint.

**Review by leseb — CHANGES_REQUESTED** [review]
Still more work to do but we are getting there, also the get_rgw_fsid() looks good now.

File: deploy/examples/create-external-cluster-resources.py (hunk 22 above)
Comment (parth-gr) [reply deploy/examples/create-external-cluster-resources.py:872]: The `info=read` PR is been backporting to pacific, I don't think so we would need to verify it then, at least for 4.10  will work fine then.
https://tracker.ceph.com/issues/51432

File: deploy/examples/create-external-cluster-resources.py (hunk 21 above)
Comment (parth-gr) [reply deploy/examples/create-external-cluster-resources.py:972]: To stop calling onto endpoint this has been added while called up from unit test, 
Otherwise for mocking we need to have something like a fake request method that does the necessity(may work),

This is the the way it is mocked in other places too, so I picked it like that.

**Review by leseb — CHANGES_REQUESTED** [review]

File: deploy/examples/create-external-cluster-resources.py [hunk 23]
```diff
@@ -766,9 +867,48 @@ def get_ceph_dashboard_link(self):
...
+                                            stderr=subprocess.PIPE)
+        except subprocess.CalledProcessError as execErr:
+            err_msg = "failed to execute command %s. Output: %s. Code: %s. Error: %s" % (
+                    cmd, execErr.output, execErr.returncode, execErr.stderr)
+            raise Exception(err_msg) 
+        output = str(output)
+        # output will be like `b'ceph version 17.2.0 (43e2e60a7559d3f46c9d53f1ca875fd499a1e35e) quincy (stable)\n'`
+        # so taking up the 15 and 16th character which is 17 in this case, so it can give us the ceph version
+        # checking with '.' because the ceph version may only have 1 digit 
+        if output[16] != '.':
```
Comment (leseb) [inline deploy/examples/create-external-cluster-resources.py:884]: Let's validate that the char is actually a valid int as well.

File: deploy/examples/create-external-cluster-resources.py [hunk 24]
```diff
@@ -791,7 +931,21 @@ def create_rgw_admin_ops_user(self):
                 err_msg = "failed to execute command %s. Output: %s. Code: %s. Error: %s" % (
                     cmd, execErr.output, execErr.returncode, execErr.stderr)
                 raise Exception(err_msg)
-
+        
+        # ceph version should atleast quincy to validate rgw-endpoint
```
Comment (leseb) [inline deploy/examples/create-external-cluster-resources.py:935]: Actually, I just realized... There is one problem though, downstream has support for this backported and the upstream backport is still pending... So this will be skipped downstream :( 
Can we read the output error from radosgw-admin when failing with unknown caps?

Comment (parth-gr) [reply deploy/examples/create-external-cluster-resources.py]: Done

File: deploy/examples/create-external-cluster-resources.py [hunk 25]
```diff
@@ -791,9 +892,26 @@ def create_rgw_admin_ops_user(self):
...
+        # separately add info=read caps, because sometimes users already exited and the cap doesn't update
+        info_cap_supported = True
+        cmd = ['radosgw-admin', 'caps', 'add', '--uid', self.EXTERNAL_RGW_ADMIN_OPS_USER_NAME,
+            '--caps', 'info=read']
+        try:
+            output = subprocess.check_output(cmd,
+                                            stderr=subprocess.PIPE)
+        except subprocess.CalledProcessError as execErr:
+            # if the ceph version not supported for adding `info=read` cap(rgw_validation)
+            if execErr.stderr.decode("utf-8") == 'could not add caps: unable to add caps: info=read\n' and execErr.returncode == 244:
```
Comment (leseb) [inline deploy/examples/create-external-cluster-resources.py:906]: can we do something like `strings.Contains` instead?

Comment (parth-gr) [reply deploy/examples/create-external-cluster-resources.py]: Actually, it works perfectly as the canary test is running on pacific and I tested with it.

File: deploy/examples/create-external-cluster-resources.py [hunk 26]
```diff
@@ -826,7 +944,46 @@ def validate_rados_namespace(self):
...
+        request_url = base_url + urllib.parse.urlencode(params)
+            
+        try:
+            r = requests.get(request_url, auth=S3Auth(access_key, secret_key,rgw_endpoint), cert=cert, verify=verify)
+        except requests.exceptions.Timeout:
+            raise ExecutionFailureException(
+                    "invalid endpoint:, not able to call admin-ops api{}".format(rgw_endpoint))
+        r1 = r.json()
+        if r1 is None or r1.get('info') is None:
+            return "NULL"
```
Comment (leseb) [inline deploy/examples/create-external-cluster-resources.py:971]: Let's return an exception for this case? or just an empty string?

**Review by leseb — CHANGES_REQUESTED** [review]
Final request :)

File: deploy/examples/create-external-cluster-resources.py (hunk 26 above)
Comment (parth-gr) [reply deploy/examples/create-external-cluster-resources.py]: Returned empty string, as an exception will be thrown by validate_rgw_endpoint

**Review by leseb — APPROVED** [review]
